File: vripper-server/src/main/java/tn/mnlr/vripper/host/ImxHost.java
package tn.mnlr.vripper.host;

import lombok.extern.slf4j.Slf4j;
import org.apache.http.NameValuePair;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.protocol.HttpClientContext;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import tn.mnlr.vripper.exception.HostException;
import tn.mnlr.vripper.exception.HtmlProcessorException;
import tn.mnlr.vripper.exception.XpathException;
import tn.mnlr.vripper.services.ConnectionService;
import tn.mnlr.vripper.services.HostService;
import tn.mnlr.vripper.services.HtmlProcessorService;
import tn.mnlr.vripper.services.XpathService;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

@Service
@Slf4j
public class ImxHost extends Host {

  private static final String host = "imx.to";
  private static final String CONTINUE_BUTTON_XPATH = "//*[@name='imgContinue']";
  private static final String IMG_XPATH = "//img[@class='centred']";

  private final HostService hostService;
  private final XpathService xpathService;
  private final ConnectionService cm;
  private final HtmlProcessorService htmlProcessorService;

  @Autowired
  public ImxHost(
      HostService hostService,
      XpathService xpathService,
      ConnectionService cm,
      HtmlProcessorService htmlProcessorService) {
    this.hostService = hostService;
    this.xpathService = xpathService;
    this.cm = cm;
    this.htmlProcessorService = htmlProcessorService;
  }

  @Override
  public String getHost() {
    return host;
  }

  @Override
  public String getLookup() {
    return host;
  }

  @Override
  public HostService.NameUrl getNameAndUrl(final String _url, final HttpClientContext context)
      throws HostException {

    String url = _url.replace("http://", "https://");
    HostService.Response resp = hostService.getResponse(url, context);
    Document doc = resp.getDocument();

    Node contDiv;
    String value = null;
    try {
      log.debug(String.format("Looking for xpath expression %s in %s", CONTINUE_BUTTON_XPATH, url));
      contDiv = xpathService.getAsNode(doc, CONTINUE_BUTTON_XPATH);
      if (contDiv == null) {
        throw new HostException(CONTINUE_BUTTON_XPATH + " cannot be found");
      }
      Node node = contDiv.getAttributes().getNamedItem("value");
      if (node != null) {
        value = node.getTextContent();
      }
    } catch (XpathException e) {
      throw new HostException(e);
    }

    if (value == null) {
      throw new HostException("Failed to obtain value attribute from continue input");
    }
    log.debug(String.format("Click button found for %s", url));
    HttpClient client = cm.getClient().build();
    HttpPost httpPost = cm.buildHttpPost(url, context);
    List<NameValuePair> params = new ArrayList<>();
    params.add(new BasicNameValuePair("imgContinue", value));
    try {
      httpPost.setEntity(new UrlEncodedFormEntity(params));
    } catch (Exception e) {
      throw new HostException(e);
    }
    log.debug(String.format("Requesting %s", httpPost));
    try (CloseableHttpResponse response =
        (CloseableHttpResponse) client.execute(httpPost, context)) {
      log.debug(String.format("Cleaning response for %s", httpPost));
      doc = htmlProcessorService.clean(EntityUtils.toString(response.getEntity()));
      EntityUtils.consumeQuietly(response.getEntity());
    } catch (IOException | HtmlProcessorException e) {
      throw new HostException(e);
    }

    Node imgNode;
    try {
      log.debug(String.format("Looking for xpath expression %s in %s", IMG_XPATH, url));
      imgNode = xpathService.getAsNode(doc, IMG_XPATH);
    } catch (XpathException e) {
      throw new HostException(e);
    }

    try {
      log.debug(String.format("Resolving name and image url for %s", url));
      String imgTitle = imgNode.getAttributes().getNamedItem("alt").getTextContent().trim();
      String imgUrl = imgNode.getAttributes().getNamedItem("src").getTextContent().trim();

      return new HostService.NameUrl(
          imgTitle.isEmpty() ? imgUrl.substring(imgUrl.lastIndexOf('/') + 1) : imgTitle, imgUrl);
    } catch (Exception e) {
      throw new HostException("Unexpected error occurred", e);
    }
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/host/ImxHost.kt
package me.mnlr.vripper.host

import me.mnlr.vripper.delegate.LoggerDelegate
import me.mnlr.vripper.download.ImageDownloadContext
import me.mnlr.vripper.exception.HostException
import me.mnlr.vripper.exception.HtmlProcessorException
import me.mnlr.vripper.exception.XpathException
import me.mnlr.vripper.services.*
import org.apache.hc.client5.http.classic.HttpClient
import org.apache.hc.client5.http.classic.methods.HttpPost
import org.apache.hc.client5.http.entity.UrlEncodedFormEntity
import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse
import org.apache.hc.core5.http.NameValuePair
import org.apache.hc.core5.http.io.entity.EntityUtils
import org.apache.hc.core5.http.message.BasicNameValuePair
import org.w3c.dom.Document
import org.w3c.dom.Node
import java.io.IOException

class ImxHost(
    private val httpService: HTTPService,
    dataTransaction: DataTransaction,
    downloadSpeedService: DownloadSpeedService,
) : Host("imx.to", httpService, dataTransaction, downloadSpeedService) {
    private val log by LoggerDelegate()

    @Throws(HostException::class)
    override fun resolve(
        url: String,
        document: Document,
        context: ImageDownloadContext
    ): Pair<String, String> {
        val httpsUrl = url.replace("http:", "https:")
        var value: String? = null
        try {
            log.debug("Looking for xpath expression $CONTINUE_BUTTON_XPATH in $httpsUrl")
            val contDiv = XpathService.getAsNode(document, CONTINUE_BUTTON_XPATH)
                ?: throw HostException("$CONTINUE_BUTTON_XPATH cannot be found")
            val node = contDiv.attributes.getNamedItem("value")
            if (node != null) {
                value = node.textContent
            }
        } catch (e: XpathException) {
            throw HostException(e)
        }
        if (value == null) {
            throw HostException("Failed to obtain value attribute from continue input")
        }
        log.debug("Click button found for $httpsUrl")
        val client: HttpClient = httpService.clientBuilder.build()
        val httpPost: HttpPost = httpService.buildHttpPost(httpsUrl, context.httpContext)
        val params: MutableList<NameValuePair> = ArrayList()
        params.add(BasicNameValuePair("imgContinue", value))
        try {
            httpPost.entity = UrlEncodedFormEntity(params)
        } catch (e: Exception) {
            throw HostException(e)
        }
        log.debug("Requesting $httpPost")
        val doc = try {
            (client.execute(
                httpPost, context.httpContext
            ) as CloseableHttpResponse).use { response ->
                log.debug("Cleaning response for $httpPost")
                try {
                    HtmlProcessorService.clean(response.entity.content)
                } finally {
                    EntityUtils.consumeQuietly(response.entity)
                }
            }
        } catch (e: IOException) {
            throw HostException(e)
        } catch (e: HtmlProcessorException) {
            throw HostException(e)
        }
        val imgNode: Node = try {
            log.debug("Looking for xpath expression $IMG_XPATH in $httpsUrl")
            XpathService.getAsNode(doc, IMG_XPATH)
        } catch (e: XpathException) {
            throw HostException(e)
        } ?: throw HostException(
            "Xpath $IMG_XPATH cannot be found in $httpsUrl"
        )
        return try {
            log.debug("Resolving name and image url for $httpsUrl")
            val imgTitle = imgNode.attributes.getNamedItem("alt").textContent.trim { it <= ' ' }
            val imgUrl = imgNode.attributes.getNamedItem("src").textContent.trim { it <= ' ' }
            Pair(
                imgTitle.ifEmpty { getDefaultImageName(imgUrl) }, imgUrl
            )
        } catch (e: Exception) {
            throw HostException("Unexpected error occurred", e)
        }
    }

    companion object {
        private const val CONTINUE_BUTTON_XPATH = "//*[@name='imgContinue']"
        private const val IMG_XPATH = "//img[@class='centred']"
    }
}